File: crates/qnero-wallet/src/select.rs
//! Choosing which notes to spend.
//!
//! The leaf circuit has exactly two input slots, so a spend reaches at most
//! two notes and the rest of the balance is unreachable in one transaction.
//! Largest first, because it is the selection that reaches the largest payment
//! from a given wallet, and because it keeps the note count falling rather
//! than growing a tail of dust a two-input circuit can never sweep.

use anyhow::{bail, Result};

use crate::store::StoredNote;

/// Input slots in the leaf circuit.
pub const MAX_INPUTS: usize = 2;

/// Pick up to [`MAX_INPUTS`] unspent notes covering `target`, largest first.
///
/// `target` is the payment plus the fee: both leave the pool, and the balance
/// equation is `inputs = outputs + fee`.
pub fn select_notes<'a>(
    unspent: impl IntoIterator<Item = &'a StoredNote>,
    target: u64,
) -> Result<Vec<&'a StoredNote>> {
    let mut candidates: Vec<&StoredNote> = unspent.into_iter().collect();
    // Ties broken by leaf index so the choice is deterministic: two notes of
    // equal value are otherwise ordered by however the store was written, and
    // a wallet that picks differently on a retry proves a different leaf.
    candidates.sort_by(|a, b| {
        b.value
            .cmp(&a.value)
            .then_with(|| a.leaf_index.cmp(&b.leaf_index))
    });

    let mut chosen = Vec::new();
    let mut total = 0u64;
    for note in candidates.iter().take(MAX_INPUTS) {
        chosen.push(*note);
        total = total.saturating_add(note.value);
        if total >= target {
            return Ok(chosen);
        }
    }

    let reachable = total;
    let held: u64 = candidates.iter().map(|note| note.value).sum();
    if candidates.is_empty() {
        bail!("this wallet holds no unspent notes");
    }
    bail!(
        "need {target} quanta and the {} largest of {} unspent notes reach {reachable} \
         ({held} held in total). A spend has {MAX_INPUTS} input slots, so consolidate first: \
         send yourself the largest notes to merge them.",
        chosen.len(),
        candidates.len()
    );
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::store::NoteOrigin;
    use qnero_notes::Digest;

    fn note(value: u64, leaf_index: u64) -> StoredNote {
        StoredNote {
            leaf_index,
            block_number: Some(1),
            value,
            commitment: Digest::hash_bytes(&[b"cm", &leaf_index.to_le_bytes()]).to_hex(),
            nullifier: Digest::hash_bytes(&[b"nf", &leaf_index.to_le_bytes()])
                .to_hex()
                .into(),
            rho: Digest::hash_bytes(&[b"rho", &leaf_index.to_le_bytes()])
                .to_hex()
                .into(),
            r: Digest::hash_bytes(&[b"r", &leaf_index.to_le_bytes()])
                .to_hex()
                .into(),
            memo: String::new(),
            origin: NoteOrigin::Shield,
            spent: false,
            spent_seen_at_block: None,
            on_chain: true,
        }
    }

    #[test]
    fn one_note_is_enough_when_it_covers_the_target() {
        let notes = [note(1_000, 0), note(400, 1)];
        let chosen = select_notes(notes.iter(), 300).unwrap();
        assert_eq!(chosen.len(), 1);
        assert_eq!(chosen[0].value, 1_000);
    }

    #[test]
    fn a_second_note_is_added_only_when_the_first_falls_short() {
        let notes = [note(200, 0), note(400, 1), note(50, 2)];
        let chosen = select_notes(notes.iter(), 500).unwrap();
        assert_eq!(
            chosen.iter().map(|n| n.value).collect::<Vec<_>>(),
            vec![400, 200]
        );
    }

    /// The circuit has two input slots. A balance spread over three notes is
    /// not reachable in one spend, and the wallet has to say so. A witness
    /// built anyway carries a balance equation that cannot hold.
    #[test]
    fn a_balance_spread_over_three_notes_is_refused_with_the_reachable_total() {
        let notes = [note(100, 0), note(100, 1), note(100, 2)];
        let error = select_notes(notes.iter(), 250).unwrap_err().to_string();
        assert!(error.contains("reach 200"), "{error}");
        assert!(error.contains("300 held in total"), "{error}");
    }

    #[test]
    fn the_fee_is_part_of_the_target() {
        let notes = [note(300, 0)];
        assert!(select_notes(notes.iter(), 300).is_ok());
        assert!(select_notes(notes.iter(), 301).is_err());
    }

    /// Two notes of equal value must be picked in one fixed order: a retry
    /// that picked the other one would spend a different note.
    #[test]
    fn equal_values_break_ties_on_the_leaf_index() {
        let notes = [note(100, 7), note(100, 3)];
        let chosen = select_notes(notes.iter(), 100).unwrap();
        assert_eq!(chosen[0].leaf_index, 3);
    }

    #[test]
    fn an_empty_wallet_says_so() {
        let empty: Vec<StoredNote> = Vec::new();
        let error = select_notes(empty.iter(), 1).unwrap_err().to_string();
        assert!(error.contains("no unspent notes"), "{error}");
    }
}

File: wallet-web/src/wallet/select.ts
/**
 * Choosing which notes to spend.
 *
 * The leaf circuit has exactly two input slots, so a spend reaches at most two
 * notes and the rest of a balance is unreachable in one transaction. Largest
 * first, because that is the selection that reaches the largest payment from a
 * given wallet and it keeps the note count falling rather than growing a tail
 * of dust a two-input circuit can never sweep.
 *
 * Ties break on the lowest leaf index. Two notes of equal value are otherwise
 * ordered by however the store was written, and a wallet that picked
 * differently on a retry would prove a different leaf and pay for a second
 * proof.
 *
 * This mirrors `crates/qnero-wallet/src/select.rs`, including the message,
 * because "consolidate first" is the only thing a person can do about it.
 */

import type { NoteRow, StoredNote } from './model';

/** Input slots in the leaf circuit. */
export const MAX_INPUTS = 2;

/**
 * The notes a spend may choose from.
 *
 * On chain, unspent, and one member per nullifier: a private batch constrains
 * its two nullifiers pairwise distinct, so two members of a conflict set in
 * one leaf is a proof that fails in circuit. The member a spend uses is the
 * largest, ties on the lowest leaf index, which is the same rule that counts
 * the set once in a balance.
 */
export function spendable(
  notes: readonly StoredNote[],
  nullifierOf: (note: StoredNote) => string,
): StoredNote[] {
  const best = new Map<string, StoredNote>();
  for (const note of notes) {
    if (note.spent || !note.onChain) {
      continue;
    }
    const nullifier = nullifierOf(note);
    const held = best.get(nullifier);
    if (held === undefined || beats(note, held)) {
      best.set(nullifier, note);
    }
  }
  return [...best.values()];
}

function beats(candidate: StoredNote, held: StoredNote): boolean {
  const a = BigInt(candidate.value);
  const b = BigInt(held.value);
  if (a !== b) {
    return a > b;
  }
  return candidate.leafIndex < held.leafIndex;
}

export interface Selection {
  notes: StoredNote[];
  total: bigint;
}

/**
 * Up to [`MAX_INPUTS`] notes covering `target`, largest first.
 *
 * `target` is the payment plus the fee: both leave the pool, and the balance
 * equation the circuit enforces is `inputs = outputs + fee`.
 */
export function selectNotes(candidates: readonly StoredNote[], target: bigint): Selection {
  const sorted = [...candidates].sort((a, b) => {
    const valueA = BigInt(a.value);
    const valueB = BigInt(b.value);
    if (valueA !== valueB) {
      return valueA > valueB ? -1 : 1;
    }
    return a.leafIndex - b.leafIndex;
  });

  const chosen: StoredNote[] = [];
  let total = 0n;
  for (const note of sorted.slice(0, MAX_INPUTS)) {
    chosen.push(note);
    total += BigInt(note.value);
    if (total >= target) {
      return { notes: chosen, total };
    }
  }

  if (sorted.length === 0) {
    throw new Error('this wallet holds no unspent notes');
  }
  const held = sorted.reduce((sum, note) => sum + BigInt(note.value), 0n);
  throw new Error(
    `this spend needs ${target} quanta and the ${chosen.length} largest of ${sorted.length} ` +
      `unspent notes reach ${total} (${held} held in total). A spend has ${MAX_INPUTS} input ` +
      'slots, so consolidate first: send yourself the largest notes to merge them.',
  );
}

/** What one spend can reach: the two largest spendable notes, and no more. */
export function reachableTotal(candidates: readonly StoredNote[]): bigint {
  return [...candidates]
    .sort((a, b) => (BigInt(b.value) > BigInt(a.value) ? 1 : -1))
    .slice(0, MAX_INPUTS)
    .reduce((sum, note) => sum + BigInt(note.value), 0n);
}

/**
 * One row per nullifier, for a table a reader adds up.
 *
 * A sender chooses `rho` and `r`, so a repeated pair yields two notes sharing
 * one nullifier of which at most one can ever settle. Both are held, because
 * refusing the second on arrival decides by arrival order and the sender
 * controls that. The balance headings count the set once, at the member a
 * spend would use, and the table under them printed every member with its own
 * amount: a reader adding the rows got a number the chain will never back.
 *
 * The surviving member is the one a spend would choose, which is the rule
 * `spendable` uses and the rule `crates/qnero-wallet/src/store.rs` collapses
 * on: a member a spend could use outranks one it could not, then the larger
 * value, then the lower leaf index. The count of members stays on the row it
 * survives as, so the conflict is still visible.
 *
 * A locked wallet opens no secret, so the key falls back to the commitment and
 * nothing collapses. That is the same fallback the balance uses, and the two
 * agree in that state as well.
 */
export function collapseRows(rows: readonly NoteRow[]): NoteRow[] {
  const best = new Map<string, NoteRow>();
  for (const row of rows) {
    const key = row.secret?.nullifier ?? row.note.commitment;
    const held = best.get(key);
    if (held === undefined || outranksForDisplay(row.note, held.note)) {
      // A `Map` keeps the position a key was first inserted at, so re-setting
      // one keeps the table's order stable as the winner changes.
      best.set(key, row);
    }
  }
  return [...best.values()];
}

function outranksForDisplay(candidate: StoredNote, held: StoredNote): boolean {
  const candidateUsable = candidate.onChain && !candidate.spent;
  const heldUsable = held.onChain && !held.spent;
  if (candidateUsable !== heldUsable) {
    return candidateUsable;
  }
  return beats(candidate, held);
}
